### ISSUE TITLE
gloo backend stops working when build with DGLOO_USE_GPUDIRECT

### ISSUE THREAD

**Issue by ybzhou** [pytorch/pytorch#2835]
this is related to merged pull request #2827, when install pytorch with DGLOO_USE_GPUDIRECT enabled, got the following error when using `distributed.all_reduce`

```
File "/home/ubuntu/anaconda3/lib/python3.6/site-packages/torch/distributed/__init__.py", line 212, in all_reduce
    return torch._C._dist_all_reduce(tensor, op, group)
RuntimeError: [enforce fail at /home/ubuntu/git/pytorch/torch/lib/gloo/gloo/common/linux.cc:103] rv != -1. -1 vs -1. Invalid argument
```

**Comment by pietern** [comment]
This is a problem in Gloo. The error can be more descriptive though... I think this is caused by assuming sysfs is available at /sys. More specifically it tries to access PCI device information at `/sys/bus/pci/devices/`. Do you have that path mounted?

**Comment by teng-li** [comment]
I think this option should only be enabled if you are using Inifiniband, but Pytorch currently uses Gloo TCP by default.

**Comment by ybzhou** [comment]
@pietern yeah, the path is mounted

**Comment by ybzhou** [comment]
@teng-li I see, I am using AWS, I guess the instance does not have Inifiniband. I am, however, using distributed on a single machine, where gpudirect should work, isn't it?

**Comment by pietern** [comment]
No, it only works with InfiniBand. @teng-li I think we need an assert for `context->getDevice()->hasGPUDirect()` at runtime so we can catch this case early. Also added an issue to check for this in Gloo.

**Comment by teng-li** [comment]
@pietern Yeah, that would really good so that we can catch it early, thanks! I created another pull request to dynamically determine if we should enable GPU Direct

**Comment by soumith** [comment]
fixed via #2841
